Better error message for old versions of Git

Sudha247 opened the pull request:
Old versions of git don't support the `--no-write-fetch-head` flag. This patch detects this particular failure and prints an error message hinting that. Fixes #10976. Prints the following error message --

```
Error: Your git version doesn't support the '--no-write-fetch-head' flag.
Hint: Please update your git version.
```

maiste commented on boot/libs.ml:
This file should be removed from your commits.

maiste commented on src/dune_pkg/rev_store.ml:
```diff
@@ -168,14 +168,34 @@ end
+      User_error.raise
+        [ Pp.text "Your git version doesn't support the '--no-write-fetch-head' flag." ]
+        ~hints:[ Pp.text "Please update your git version." ]
+    | _ ->
+      ();
+      Error { Git_error.dir; args; exit_code; output = [] })
```
```suggestion
    | _ -> Error { Git_error.dir; args; exit_code; output = [] })
```
Maybe I misunderstand something: why do you need `()`?

Sudha247 replied:
Missed it during refactoring a previous attempt. Removed now. Nice spot!

Sudha247 committed:
Better error message for old versions of Git

Old versions of git don't support the  flag. This patch detects this particular failure and prints an error message hinting that.

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

maiste commented on src/dune_rules/setup.defaults.ml:
```diff
@@ -11,7 +11,7 @@ let roots : string option Install.Roots.t =
   ; libexec_root = None
   }
 
+let prefix : string option = None
```
Is it necessary to add the type here? I don't see where it is used in the code above.

Leonidas-from-XIV replied:
This seems to be an unrelated change that happens due to the changes from #11172. I've created a PR to commit it separately: #11198.

Leonidas-from-XIV reviewed (changes requested):
Nice. I like that it doesn't add additional calls to detect that flag!

Added some comments to improve and probably needs to either be rebased on #11198 when it is merged or just exclude the change to `setup.default.ml`.

Sudha247 committed:
Address review comments

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Sudha247 commented on src/dune_pkg/rev_store.ml:
```diff
@@ -168,14 +168,34 @@ end
+      when String.is_prefix ~prefix:"error: unknown option `no-write-fetch-head'" stderr
+      ->
+      User_error.raise
+        [ Pp.text "Your git version doesn't support the '--no-write-fetch-head' flag." ]
```
Makes sense, done.

Sudha247 commented on src/dune_pkg/rev_store.ml:
```diff
@@ -168,14 +168,34 @@ end
+          Stdune.Io.read_file path, exit_code)
+      ~prefix:"dune"
+      ~suffix:"run_with_exit_code"
+      ~dir:(Path.of_string (Filename.get_temp_dir_name ()))
```
As far as I can tell, it'd be computed the same number of times if it were in the function definition - I checked it with some debug prints, but maybe I'm missing something.

maiste reviewed (approved):
LGTM :+1: 
(CI issue is unrelated to this PR)

rgrinberg committed:
refactor: some cosmetic tweaks

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
fix(pkg): print stderr on git errors

instead of discarding the stderr

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>
